src/main/java/graphql/analysis/QueryTransformer.java

```diff
@@ -0,0 +1,172 @@
+package graphql.analysis;
+
+import graphql.PublicApi;
+import graphql.language.FragmentDefinition;
+import graphql.language.Node;
+import graphql.language.NodeTraverser.LeaveOrEnter;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLSchema;
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+import graphql.util.TraverserVisitor;
+import graphql.util.TreeTransformer;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static graphql.Assert.assertNotNull;
+import static graphql.language.AstNodeAdapter.AST_NODE_ADAPTER;
+
+/**
+ * Helps to transform a Document (or parts of it) and tracks at the same time the corresponding Schema types.
+ * <p>
+ * This is an important distinction to just traversing the Document without any type information: Each field has a clearly
+ * defined type. See {@link QueryVisitorFieldEnvironment}.
+ * <p>
+ * Furthermore are the built in Directives skip/include automatically evaluated: if parts of the Document should be ignored they will not
+ * be visited. But this is not a full evaluation of a Query: every fragment will be visited/followed regardless of the type condition.
+ * <p>
+ * It also doesn't consider field merging, which means for example {@code { user{firstName} user{firstName}} } will result in four
+ * visitField calls.
+ */
+@PublicApi
+public class QueryTransformer {
+
+    private final Node root;
+    private final GraphQLSchema schema;
+    private final Map<String, FragmentDefinition> fragmentsByName;
+    private final Map<String, Object> variables;
+
+    private final GraphQLObjectType rootParentType;
+
+
+    private QueryTransformer(GraphQLSchema schema,
+                             Node root,
+                             GraphQLObjectType rootParentType,
+                             Map<String, FragmentDefinition> fragmentsByName,
+                             Map<String, Object> variables) {
+        this.schema = assertNotNull(schema, "schema can't be null");
+        this.variables = assertNotNull(variables, "variables can't be null");
+        this.root = assertNotNull(root, "root can't be null");
+        this.rootParentType = assertNotNull(rootParentType);
+        this.fragmentsByName = assertNotNull(fragmentsByName, "fragmentsByName can't be null");
+    }
+
+    /**
+     * Visits the Document in pre-order and allows to transform it using {@link graphql.util.TreeTransformerUtil}
+     * methods. Please note that fragment spreads are not followed and need to be
+     * processed explicitly by supplying them as root.
+     *
+     * @param queryVisitor the query visitor that will be called back.
+     *
+     * @return changed root
+     *
+     * @throws IllegalArgumentException if there are multiple root nodes.
+     */
+    public Node transform(QueryVisitor queryVisitor) {
```

Comment:
Are we going to extract this method as well somewhere else? Also one thing that is missing in QueryVisitor that we need in Nadel is being able to visit FragmentDefinition (needed so that we can change type condition on the fragment)